File: dotnet/src/Rerout/Resources/Tags.cs
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Rerout.Internal;
using Rerout.Models;

namespace Rerout.Resources;

/// <summary>
/// Tag management: list, create, update, delete. Reached through
/// <see cref="ReroutClient.Tags"/>. The project is resolved from the API key —
/// there is no project id in the path.
/// </summary>
public sealed class Tags
{
    private readonly ReroutHttpHandler _http;

    internal Tags(ReroutHttpHandler http) => _http = http;

    /// <summary>
    /// List the project's tags with their live (non-deleted) link counts.
    /// </summary>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>The tags registered to the project, each as a <see cref="TagSummary"/>.</returns>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<ListTagsResult> ListAsync(CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<ListTagsResult>(
            new ReroutRequest
            {
                Method = HttpMethod.Get,
                Path = "/v1/projects/me/tags",
            },
            cancellationToken);
    }

    /// <summary>
    /// Create a tag. <see cref="CreateTagInput.Color"/> is optional; the server
    /// validates it and defaults to <c>teal</c> when omitted.
    /// </summary>
    /// <param name="input">The tag to create. <c>Name</c> is required.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>The created tag (without a link count).</returns>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<Tag> CreateAsync(CreateTagInput input, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<Tag>(
            new ReroutRequest
            {
                Method = HttpMethod.Post,
                Path = "/v1/projects/me/tags",
                Body = input,
            },
            cancellationToken);
    }

    /// <summary>
    /// Update a tag's name and/or color. Only the fields set on
    /// <paramref name="input"/> are sent; omitted fields are left unchanged.
    /// </summary>
    /// <param name="tagId">The tag identifier (<c>tag_…</c>).</param>
    /// <param name="input">The fields to change. At least one must be set.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>The updated tag (without a link count).</returns>
    /// <exception cref="ReroutException">
    /// Thrown client-side with code <c>empty_update</c> when
    /// <paramref name="input"/> has no fields to send — the request never reaches
    /// the API. (Mirrors <see cref="Links.UpdateAsync"/>.)
    /// </exception>
    public Task<Tag> UpdateAsync(string tagId, UpdateTagInput input, CancellationToken cancellationToken = default)
    {
        if (input.IsEmpty)
        {
            throw new ReroutException(
                "empty_update",
                "UpdateTagInput has no fields to send. Set at least one field.",
                0,
                path: $"/v1/projects/me/tags/{tagId}");
        }

        return _http.SendAsync<Tag>(
            new ReroutRequest
            {
                Method = HttpMethod.Patch,
                Path = $"/v1/projects/me/tags/{EncodeId(tagId)}",
                Body = input,
            },
            cancellationToken);
    }

    /// <summary>
    /// Delete a tag and drop its assignments from all links. Idempotent.
    /// </summary>
    /// <param name="tagId">The tag identifier (<c>tag_…</c>).</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>A <see cref="DeleteResult"/> with <c>Deleted = true</c>.</returns>
    /// <exception cref="ReroutException">On any API or transport failure.</exception>
    public Task<DeleteResult> DeleteAsync(string tagId, CancellationToken cancellationToken = default)
    {
        return _http.SendAsync<DeleteResult>(
            new ReroutRequest
            {
                Method = HttpMethod.Delete,
                Path = $"/v1/projects/me/tags/{EncodeId(tagId)}",
            },
            cancellationToken);
    }

    internal static string EncodeId(string tagId) => Uri.EscapeDataString(tagId);
}

File: go/rerout/tags.go
// Package rerout — Tags management namespace.
//
// Reached via Client.Tags(). Wraps the /v1/projects/me/tags surface: list,
// create, update, delete. The project is resolved from the API key, so there
// is no project id in the path.

package rerout

import (
	"context"
	"net/url"
)

// Tags is the tag-management namespace. Construct via Client.Tags().
type Tags struct {
	client *Client
}

// List returns the project's tags with their live link counts. Each returned
// TagSummary carries the number of non-deleted links the tag is attached to.
func (t *Tags) List(ctx context.Context) (*ListTagsResult, error) {
	var out ListTagsResult
	err := t.client.do(ctx, requestOptions{
		method: "GET",
		path:   "/v1/projects/me/tags",
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Create creates a tag. Name is required; Color is optional and defaults
// server-side when omitted. The response omits the link count.
func (t *Tags) Create(ctx context.Context, input CreateTagInput) (*Tag, error) {
	var out Tag
	err := t.client.do(ctx, requestOptions{
		method: "POST",
		path:   "/v1/projects/me/tags",
		body:   input,
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Update patches a tag's name and/or color. Only fields set on input are sent;
// omitted fields are left unchanged. Mirrors Links.Update: an empty
// UpdateTagInput is a client-side error and never hits the API.
func (t *Tags) Update(ctx context.Context, tagID string, input UpdateTagInput) (*Tag, error) {
	if input.IsEmpty() {
		return nil, &ReroutError{
			Code:    CodeBadRequest,
			Status:  0,
			Message: "rerout: UpdateTagInput has no fields to send.",
			Path:    "/v1/projects/me/tags/" + url.PathEscape(tagID),
		}
	}
	var out Tag
	err := t.client.do(ctx, requestOptions{
		method: "PATCH",
		path:   "/v1/projects/me/tags/" + url.PathEscape(tagID),
		body:   input,
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes a tag and drops its assignments from all links. Tag ids look
// like "tag_…". The response body is {"deleted": true}.
func (t *Tags) Delete(ctx context.Context, tagID string) (*DeleteResult, error) {
	var out DeleteResult
	err := t.client.do(ctx, requestOptions{
		method: "DELETE",
		path:   "/v1/projects/me/tags/" + url.PathEscape(tagID),
	}, &out)
	if err != nil {
		return nil, err
	}
	return &out, nil
}
